Update target group destroyer to wait for listener

paladin-devops opened the pull request:
This PR updates the ECS platform destroyer to wait for there to be zero AWS ALB listeners targeting the target group to attempt to delete the target group, for a maximum of 5 minutes.

This does not fully address #4461, because my destroy attempts are still failing on deleting the security group, but it's a start! And deleting that will be fixed in a subsequent PR.

paladin-devops committed:
ecs: Update target group destroyer to wait for listener.

This commit updates the ECS platform destroyer to wait for there to be zero AWS ALB listeners targeting the target group to attempt to delete the target group, for a maximum of 5 minutes.

paladin-devops commented:
When #4496 is addressed, it might be nicer to update the ECS default releaser to modify the listener to no longer route to the target group; but this will do for now.

paladin-devops committed:
fix: ECS ALB ARN error format.

briancain commented on builtin/aws/ecs/platform.go:
```diff
@@ -1339,21 +1339,89 @@ func (p *Platform) resourceTargetGroupDestroy(
 	}
 
-	s := sg.Add("Deleting target group %s", state.Name)
+	s := sg.Add("Getting target group details", state.Name)
```
Might be nice to preserve the detail about the target group name here?

briancain commented on builtin/aws/ecs/platform.go:
```diff
@@ -1339,21 +1339,89 @@ func (p *Platform) resourceTargetGroupDestroy(
+
+	// There should be only one target group we "loop through" here
+	for _, group := range groups.TargetGroups {
+		s = sg.Add("Checking if target group " + *group.TargetGroupName + " has an active ALB listener")
```
I think having `*group.TargetGroupName` as a `%q` rather than the string addition makes more sense here

briancain commented on builtin/aws/ecs/platform.go:
```diff
@@ -1339,21 +1339,89 @@ func (p *Platform) resourceTargetGroupDestroy(
+	s.Done()
+
+	// There should be only one target group we "loop through" here
+	for _, group := range groups.TargetGroups {
```
This is quite a big loop! I think it's worth considering if there's any way we could reduce this down from 5 nested for loops 😅

paladin-devops committed:
ecs: Simplify target group deletion loop.

There should be only one target group returned by the ARN lookup, so this is now enforced by returning an error if that is not the case. This allows for the elimination of the most exterior loop in the deletion of the target group.
